packages/dataviews/src/hooks/use-form-validity.ts

```diff
@@ -0,0 +1,302 @@
+/**
+ * External dependencies
+ */
+import deepMerge from 'deepmerge';
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useEffect, useRef, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import normalizeFields from '../utils/normalize-fields';
+import type { Field, Form, FormValidity } from '../types';
+
+const isEmptyNullOrUndefined = ( value: any ) =>
+	[ undefined, '', null ].includes( value );
+
+const isArrayOrElementsEmptyNullOrUndefined = ( value: any ) => {
+	return (
+		! Array.isArray( value ) ||
+		value.length === 0 ||
+		value.every( ( element: any ) => isEmptyNullOrUndefined( element ) )
+	);
+};
+
+function isInvalidForRequired( fieldType: string | undefined, value: any ) {
+	if (
+		( fieldType === undefined && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'text' && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'email' && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'url' && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'telephone' && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'password' && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'integer' && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'number' && isEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'array' &&
+			isArrayOrElementsEmptyNullOrUndefined( value ) ) ||
+		( fieldType === 'boolean' && value !== true )
+	) {
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * Hook that validates a form item and returns an object with error messages for each field.
+ *
+ * @param item   The item to validate.
+ * @param fields Fields config.
+ * @param form   Form config.
+ *
+ * @return Record of field IDs to error messages (undefined means no error).
+ */
+export function useFormValidity< Item >(
+	item: Item,
+	fields: Field< Item >[],
+	form: Form
+): { validity: FormValidity; isValid: boolean } {
+	const [ formValidity, setFormValidity ] = useState< FormValidity >();
+
+	const previousValidatedValuesRef = useRef< Record< string, any > >( {} );
+
+	const validate = useCallback( () => {
+		if ( typeof form.fields === 'undefined' ) {
+			setFormValidity( undefined );
+			return;
+		}
+
+		const normalizedFields = normalizeFields(
+			fields.filter( ( field ) => {
+				return form?.fields?.some( ( formField ) => {
+					if ( typeof formField === 'string' ) {
+						return formField === field.id;
+					}
+
+					return formField.id === field.id;
+				} );
+			} )
+		);
+
+		normalizedFields.forEach( ( field ) => {
+			const value = field.getValue( { item } );
+			if (
+				previousValidatedValuesRef.current.hasOwnProperty( field.id ) &&
+				value === previousValidatedValuesRef.current[ field.id ]
+			) {
+				return;
+			}
+			previousValidatedValuesRef.current[ field.id ] = value;
+
+			// Check isValid.required
+			if (
+				field.isValid.required &&
```

Comment:
Personally I wonder if we should handle these things globally like here, or defer to the field types instead. Maybe for "required" it's ok to have a global handling but for most deferring to the field types seems like a better idea. (We can iterate on this)